### ISSUE TITLE
Allow client methods to dictate what HTTP status codes are errors or otherwise unexpected

### ISSUE THREAD

**Issue by heaths** [Azure/azure-sdk-for-rust#1733]
Like with other Azure SDK languages, client methods need to specify what HTTP success and failure status codes it expects, and should probably err for unexpected errors with a different error variant or at least something to denote that 1) the status code wasn't expected, and 2) there may not be a JSON error response body.

**Comment by RickWinter** [comment]
@heaths Does typespec define the errors? Where would the emitter get this information?

**Comment by heaths** [comment]
Yes, TypeSpec specs define these errors; though, often these are encapsulated by Azure Core types. I imagine TCGC gets that information raw, though, right @jhendrixMSFT? E.g., an LRO should return a 202 while a DELETE should generally return a 204 but could return a 200 in some cases e.g., Key Vault. In that case, their spec has to be explicit: https://github.com/Azure/azure-rest-api-specs/blob/64b39af620f9326da0db6e012414838453be016c/specification/keyvault/Security.KeyVault.Secrets/routes.tsp#L59 which, because it returns a model, results in a 200: https://github.com/Azure/azure-rest-api-specs/blob/64b39af620f9326da0db6e012414838453be016c/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.6-preview.2/secrets.json#L321-L325

**Comment by heaths** [comment]
Still, though, we're not 100% generated so, however we dictate which HTTP status codes equate to success, we should extend that to value-added extensions.

**Comment by heaths** [comment]
Basically, we should just ignore any unexpected 2xx - services may be authored wrong - and any 4xx or 5xx would be an error.

The retry policy would be hardcode to retry on certain HTTP status codes. Customers can override the retriable HTTP status codes, but in Go it's a complete replacement. Client devs also need to be able to override or update the list, but should be a different mechanism that doesn't require override everything that the customer would have to.

Basically, look at Go's implementation for retries; however, for the general case we just treat all 2xx as success and everything else as error. @heaths will follow up with @KrzysztofCwalina to see why .NET asserts success codes.

**Comment by heaths** [comment]
@analogrelay had a good idea in #2894 - though implemented as an extension method - that would also work here. If we added something like an `async fn success(&self, options: Option<ResponseOptions>) -> Result<Self>` to `Response<T, F: Format>` and `RawResponse`, we could take a `Vec<StatusCode>` (default basically allows the full 2xx range) of expected codes: anything else is considered an error, but this is all done from within the context of the client then - not the pipeline, which we want to keep free of non-dev errors bubbling up too soon so that pipeline policies have a chance to run on the request and response.

We discussed this in a v-team meeting and agreed on the design, so removing the `design-discussion` label.
